#
# Biostar Workflows: https://www.biostarhandbook.com/
#

# Genome accession.
ACC = KM233118

# SRR number.
SRR = SRR1972918

# Set both reads for paired end and only one for single end.
R1 = reads/SRR1972918_1.fq
R2 = reads/SRR1972918_2.fq

# The reference genome.
REF = refs/${ACC}.fa

# The name of the gff file.
GFF = refs/${ACC}.gff

# The alignment file.
BAM = bam/${SRR}.bam

# Makefile customizations.
SHELL := bash
.DELETE_ON_ERROR:
.ONESHELL:
.SHELLFLAGS := -eu -o pipefail -c
MAKEFLAGS += --warn-undefined-variables --no-print-directory

usage:
	@echo "#"
	@echo "# Align short reads to a reference genome"
	@echo "#"
	@echo "# make run"
	@echo "#"

# Download the reference genome.
genome:
	make -f src/run/genbank.mk fasta gff ACC=${ACC} REF=${REF} GFF=${GFF}

# Download the short reads.
fastq:
	make -f src/run/sra.mk run SRR=${SRR} N=2500 R1=${R1} R2=${R2}

# Perform data related tasks
data: genome fastq
	make -f src/run/bwa.mk index REF=${REF}

# Index and align the reads to the reference genome.
align:
	make -f src/run/bwa.mk run REF=${REF} R1=${R1} R2=${R2} BAM=${BAM}

# Run the complete pipeline.
run: data align

# Remove the alignment generated by the pipeline
run!:
	rm -f ${BAM}

.PHONY: usage data align run
