#
# Biostar Workflows: https://www.biostarhandbook.com/
#

# Genome accession.
ACC = KM233118

# The reference genome.
REF = refs/${ACC}.fa

# SRR number.
SRR = SRR1972918

# The read pairs.
R1 = reads/${SRR}.fq
R2 = reads/${SRR}.fq

# The alignment file.
BAM = bam/${SRR}.bam

# The variants with bcftools.
VCF = vcf/${SRR}.vcf.gz

# Number of CPUS
NCPU = 4

# Makefile customizations.
SHELL := bash
.DELETE_ON_ERROR:
.ONESHELL:
.SHELLFLAGS := -eu -o pipefail -c
MAKEFLAGS += --warn-undefined-variables --no-print-directory

usage:
	@echo "#"
	@echo "# Short read variant calling."
	@echo "#"
	@echo "#"
	@echo "# make run"
	@echo "#"

run:
	# Download the reference genome.
	make -f src/run/genbank.mk fasta gff ACC=${ACC} REF=${REF}

	# Download the short reads.
	make -f src/run/sra.mk run SRR=${SRR} N=2500 R1=${R1} R2=${R2}

	# Index the genome then align the reads to the reference genome.
	make -f src/run/bwa.mk index run R1=${R1} R2=${R2} REF=${REF} BAM=${BAM} NCPU=${NCPU}

	# Call variants with bcftools.
	make -f src/run/bcftools.mk run REF=${REF} BAM=${BAM} VCF=${VCF} NCPU=${NCPU}

# Remove the variant calls generated by the pipeline
run!:
	rm -f ${VCF}
