\documentclass[]{ctexbook}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{xltxtra,xunicode}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[b5paper,tmargin=2.5cm,bmargin=2.5cm,lmargin=3.5cm,rmargin=2.5cm]{geometry}
\usepackage[unicode=true]{hyperref}
\PassOptionsToPackage{usenames,dvipsnames}{color} % color is loaded by hyperref
\hypersetup{
            pdftitle={Biostar},
            pdfauthor={小苏},
            colorlinks=true,
            linkcolor=Maroon,
            citecolor=Blue,
            urlcolor=Blue,
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{natbib}
\bibliographystyle{apalike}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs}
% Fix footnotes in tables (requires footnote package)
\IfFileExists{footnote.sty}{\usepackage{footnote}\makesavenoteenv{long table}}{}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

% set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother

\usepackage{booktabs}
\usepackage{longtable}

\usepackage{framed,color}
\definecolor{shadecolor}{RGB}{248,248,248}

\renewcommand{\textfraction}{0.05}
\renewcommand{\topfraction}{0.8}
\renewcommand{\bottomfraction}{0.8}
\renewcommand{\floatpagefraction}{0.75}

\let\oldhref\href
\renewcommand{\href}[2]{#2\footnote{\url{#1}}}

\makeatletter
\newenvironment{kframe}{%
\medskip{}
\setlength{\fboxsep}{.8em}
 \def\at@end@of@kframe{}%
 \ifinner\ifhmode%
  \def\at@end@of@kframe{\end{minipage}}%
  \begin{minipage}{\columnwidth}%
 \fi\fi%
 \def\FrameCommand##1{\hskip\@totalleftmargin \hskip-\fboxsep
 \colorbox{shadecolor}{##1}\hskip-\fboxsep
     % There is no \\@totalrightmargin, so:
     \hskip-\linewidth \hskip-\@totalleftmargin \hskip\columnwidth}%
 \MakeFramed {\advance\hsize-\width
   \@totalleftmargin\z@ \linewidth\hsize
   \@setminipage}}%
 {\par\unskip\endMakeFramed%
 \at@end@of@kframe}
\makeatother

\makeatletter
\@ifundefined{Shaded}{
}{\renewenvironment{Shaded}{\begin{kframe}}{\end{kframe}}}
\@ifpackageloaded{fancyvrb}{%
  % https://github.com/CTeX-org/ctex-kit/issues/331
  \RecustomVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\},formatcom=\xeCJKVerbAddon}%
}{}
\makeatother

\usepackage{makeidx}
\makeindex

\urlstyle{tt}

\usepackage{amsthm}
\makeatletter
\def\thm@space@setup{%
  \thm@preskip=8pt plus 2pt minus 4pt
  \thm@postskip=\thm@preskip
}
\makeatother

\newenvironment{cols}[1][]{}{}

\newenvironment{col}[1]{\begin{minipage}{#1}\ignorespaces}{%
	\end{minipage}
	\ifhmode\unskip\fi
	\aftergroup\useignorespacesandallpars}

\def\useignorespacesandallpars#1\ignorespaces\fi{%
	#1\fi\ignorespacesandallpars}

\makeatletter
\def\ignorespacesandallpars{%
	\@ifnextchar\par
	{\expandafter\ignorespacesandallpars\@gobble}%
	{}%
}
\makeatother

\frontmatter
\usepackage{booktabs}
\usepackage{longtable}
\usepackage{array}
\usepackage{multirow}
\usepackage{wrapfig}
\usepackage{float}
\usepackage{colortbl}
\usepackage{pdflscape}
\usepackage{tabu}
\usepackage{threeparttable}
\usepackage{threeparttablex}
\usepackage[normalem]{ulem}
\usepackage{makecell}
\usepackage{xcolor}

\title{Biostar}
\author{小苏}
\date{2025-03-24}

\begin{document}
\maketitle

\thispagestyle{empty}

\begin{center}
献给……
\end{center}

\setlength{\abovedisplayskip}{-5pt}
\setlength{\abovedisplayshortskip}{-5pt}

{
\setcounter{tocdepth}{1}
\tableofcontents
}
\listoftables
\listoffigures
\chapter*{前言}\label{ux524dux8a00}


\emph{我不是代码的创作者，我只是代码的搬运工！}

\part{数据下载}\label{part-ux6570ux636eux4e0bux8f7d}

\chapter{aria2}\label{aria2}

aria2 是一个轻量级的多协议和多源命令行下载工具。它支持 HTTP/HTTPS、FTP、BitTorrent 和 Metalink。aria2 可以通过最大化网络带宽利用率来加快下载速度。它支持 HTTP/HTTPS 代理，SOCKS 代理，HTTP 代理隧道，NAT 穿透，IPv6 和 IP 版本选择。aria2 可以通过 JSON-RPC 和 XML-RPC 接口进行控制。

官方教程：\url{https://github.com/aria2/aria2}

使用示范

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{URL}\OperatorTok{=}\NormalTok{https://ftp.ensembl.org/pub/release{-}113/fasta/homo\_sapiens/dna/Homo\_sapiens.GRCh38.dna.toplevel.fa.gz}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ src/run/aria.mk URL=}\VariableTok{$\{URL\}}\NormalTok{ run}
\end{Highlighting}
\end{Shaded}

\href{src/run/aria.mk}{aria.mk}

\chapter{curl}\label{curl}

curl 是一个命令行工具和库，用于传输数据，支持多种协议，包括 HTTP、HTTPS、FTP、FTPS、SFTP、IMAP、SMTP、POP3、LDAP、RTMP 和 RTSP。curl 还支持 SSL 证书、HTTP POST、HTTP PUT、FTP 上传、HTTP 基本身份验证、代理、cookie、用户代理、压缩、断点续传、文件传输限速、重定向等功能。

使用示范:

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{URL}\OperatorTok{=}\NormalTok{https://ftp.ensembl.org/pub/release{-}113/fasta/homo\_sapiens/dna/Homo\_sapiens.GRCh38.dna.toplevel.fa.gz}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ src/run/curl.mk URL=}\VariableTok{$\{URL\}}\NormalTok{ run}
\end{Highlighting}
\end{Shaded}

\href{src/run/curl.mk}{aria.mk}

\chapter{SRA}\label{sra}

SRA（Sequence Read Archive） 是一个由美国国家生物技术信息中心（NCBI）维护的高通量测序数据公共仓库。它是国际核苷酸序列数据库合作（INSDC）的组成部分，与欧洲生物信息研究所（EBI）和日本DNA数据库（DDBJ）共享数据。从数据库中下载原始测序数据是必备技能。

使用示范:
1. 使用 fastq-dump 工具。

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{SRR}\OperatorTok{=}\NormalTok{SRR12351448}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ src/run/sra.mk SRR=}\VariableTok{$\{SRR\}}\NormalTok{ N=ALL run}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{design}\OperatorTok{=}\NormalTok{design.csv}

\FunctionTok{cat} \VariableTok{$\{design\}} \KeywordTok{|} \ExtensionTok{parallel} \AttributeTok{{-}v} \AttributeTok{{-}{-}eta} \AttributeTok{{-}{-}lb} \AttributeTok{{-}{-}header}\NormalTok{ : }\AttributeTok{{-}{-}colsep}\NormalTok{ , }\DataTypeTok{\textbackslash{}}
\NormalTok{                make }\AttributeTok{{-}f}\NormalTok{ src/run/sra.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{                SRR=\{Run\} }\DataTypeTok{\textbackslash{}}
\NormalTok{                N=ALL }\DataTypeTok{\textbackslash{}}
\NormalTok{                run}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  使用 aria2 工具。
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{SRR}\OperatorTok{=}\NormalTok{SRR12351448}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ src/run/sra.mk SRR=}\VariableTok{$\{SRR\}}\NormalTok{ N=ALL aria}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{design}\OperatorTok{=}\NormalTok{design.csv}

\FunctionTok{cat} \VariableTok{$\{design\}} \KeywordTok{|} \ExtensionTok{parallel} \AttributeTok{{-}v} \AttributeTok{{-}{-}eta} \AttributeTok{{-}{-}lb} \AttributeTok{{-}{-}header}\NormalTok{ : }\AttributeTok{{-}{-}colsep}\NormalTok{ , }\DataTypeTok{\textbackslash{}}
\NormalTok{                make }\AttributeTok{{-}f}\NormalTok{ src/run/sra.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{                SRR=\{Run\} }\DataTypeTok{\textbackslash{}}
\NormalTok{                N=ALL }\DataTypeTok{\textbackslash{}}
\NormalTok{                aria}
\end{Highlighting}
\end{Shaded}

\href{src/run/sra.mk}{sra.mk}

\part{fastq 质量控制}\label{part-fastq-ux8d28ux91cfux63a7ux5236}

\chapter{fastp}\label{fastp}

官方教程：\url{https://github.com/OpenGene/fastp}

使用示范:

单端数据

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{SRR}\OperatorTok{=}\NormalTok{SRR12351448}
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{READS}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ src/run/fastp.mk P1=}\VariableTok{$\{READS\}}\NormalTok{/}\VariableTok{$\{SRR\}}\NormalTok{\_1.fastq run}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{READS}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}
\VariableTok{design}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/design.csv}

\FunctionTok{cat} \VariableTok{$\{design\}} \KeywordTok{|} \ExtensionTok{parallel} \AttributeTok{{-}{-}header}\NormalTok{ : }\AttributeTok{{-}{-}colsep}\NormalTok{ , }\DataTypeTok{\textbackslash{}}
\NormalTok{        make }\AttributeTok{{-}f}\NormalTok{ src/run/fastp.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{        P1=}\VariableTok{$\{READS\}}\NormalTok{/\{Run\}\_1.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        run}
\end{Highlighting}
\end{Shaded}

双端数据

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{SRR}\OperatorTok{=}\NormalTok{SRR12351448}
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{READS}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ src/run/fastp.mk P1=}\VariableTok{$\{READS\}}\NormalTok{/}\VariableTok{$\{SRR\}}\NormalTok{\_1.fastq P2=}\VariableTok{$\{READS\}}\NormalTok{/}\VariableTok{$\{SRR\}}\NormalTok{\_2.fastq run}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{READS}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}
\VariableTok{design}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/design.csv}

\FunctionTok{cat} \VariableTok{$\{design\}} \KeywordTok{|} \ExtensionTok{parallel} \AttributeTok{{-}{-}header}\NormalTok{ : }\AttributeTok{{-}{-}colsep}\NormalTok{ , }\DataTypeTok{\textbackslash{}}
\NormalTok{        make }\AttributeTok{{-}f}\NormalTok{ src/run/fastp.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{        P1=}\VariableTok{$\{READS\}}\NormalTok{/\{Run\}\_1.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        P2=}\VariableTok{$\{READS\}}\NormalTok{/\{Run\}\_2.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        run}
\end{Highlighting}
\end{Shaded}

\href{src/run/fastp.mk}{sra.mk}

\part{比对}\label{part-ux6bd4ux5bf9}

\chapter{bwa}\label{bwa}

官方教程：\url{https://github.com/lh3/bwa}

\section{构建索引}\label{ux6784ux5efaux7d22ux5f15}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{REF}\OperatorTok{=}\NormalTok{\textasciitilde{}/database/Human/Homo\_sapiens.GRCh38.dna.toplevel.fa}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ \textasciitilde{}/src/run/bwa.mk REF=}\VariableTok{$\{REF\}}\NormalTok{ index}
\end{Highlighting}
\end{Shaded}

\section{比对}\label{ux6bd4ux5bf9}

单端数据

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{REF}\OperatorTok{=}\NormalTok{genome/Human/Homo\_sapiens.GRCh38.dna.primary\_assembly.fa}
\VariableTok{BWA}\OperatorTok{=}\NormalTok{results/bwa}
\VariableTok{CLEANDATA}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}
\VariableTok{Run}\OperatorTok{=}\NormalTok{SRR1343245}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ src/run/bwa.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{        REF=}\VariableTok{$\{REF\}} \DataTypeTok{\textbackslash{}}
\NormalTok{        R1=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/}\VariableTok{$\{Run\}}\NormalTok{\_1.trimmed.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        BAM=}\VariableTok{$\{BWA\}}\NormalTok{/\{sample\}.bam }\DataTypeTok{\textbackslash{}}
\NormalTok{        run}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{design}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/design.csv}
\VariableTok{REF}\OperatorTok{=}\NormalTok{genome/Human/Homo\_sapiens.GRCh38.dna.primary\_assembly.fa}
\VariableTok{CLEANDATA}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}
\VariableTok{BWA}\OperatorTok{=}\NormalTok{results/bwa}
                
\FunctionTok{cat} \VariableTok{$\{design\}} \KeywordTok{|} \ExtensionTok{parallel} \AttributeTok{{-}{-}header}\NormalTok{ : }\AttributeTok{{-}{-}colsep}\NormalTok{ , }\DataTypeTok{\textbackslash{}}
\NormalTok{        make }\AttributeTok{{-}f}\NormalTok{ src/run/bwa.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{        REF=}\VariableTok{$\{REF\}} \DataTypeTok{\textbackslash{}}
\NormalTok{        R1=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/\{Run\}\_1.trimmed.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        BAM=}\VariableTok{$\{BWA\}}\NormalTok{/\{sample\}.bam }\DataTypeTok{\textbackslash{}}
\NormalTok{        run}
\end{Highlighting}
\end{Shaded}

双端数据

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{REF}\OperatorTok{=}\NormalTok{genome/Human/Homo\_sapiens.GRCh38.dna.primary\_assembly.fa}
\VariableTok{BWA}\OperatorTok{=}\NormalTok{results/bwa}
\VariableTok{CLEANDATA}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}
\VariableTok{Run}\OperatorTok{=}\NormalTok{SRR1343245}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ src/run/bwa.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{        REF=}\VariableTok{$\{REF\}} \DataTypeTok{\textbackslash{}}
\NormalTok{        R1=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/}\VariableTok{$\{Run\}}\NormalTok{\_1.trimmed.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        R2=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/}\VariableTok{$\{Run\}}\NormalTok{\_2.trimmed.fastq}
        \VariableTok{BAM}\OperatorTok{=}\VariableTok{$\{BWA\}}\NormalTok{/\{sample\}.bam }\DataTypeTok{\textbackslash{}}
        \ExtensionTok{run}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{design}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/design.csv}
\VariableTok{REF}\OperatorTok{=}\NormalTok{genome/Human/Homo\_sapiens.GRCh38.dna.primary\_assembly.fa}
\VariableTok{CLEANDATA}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}
\VariableTok{BWA}\OperatorTok{=}\NormalTok{results/bwa}
                
\FunctionTok{cat} \VariableTok{$\{design\}} \KeywordTok{|} \ExtensionTok{parallel} \AttributeTok{{-}{-}header}\NormalTok{ : }\AttributeTok{{-}{-}colsep}\NormalTok{ , }\DataTypeTok{\textbackslash{}}
\NormalTok{        make }\AttributeTok{{-}f}\NormalTok{ src/run/bwa.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{        REF=}\VariableTok{$\{REF\}} \DataTypeTok{\textbackslash{}}
\NormalTok{        R1=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/\{Run\}\_1.trimmed.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        R2=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/\{Run\}\_2.trimmed.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        BAM=}\VariableTok{$\{BWA\}}\NormalTok{/\{sample\}.bam }\DataTypeTok{\textbackslash{}}
\NormalTok{        run}
\end{Highlighting}
\end{Shaded}

\href{src/run/bwa.mk}{sra.mk}

\chapter{hisat2}\label{hisat2}

官方教程：\url{https://daehwankimlab.github.io/hisat2/}

\section{构建索引}\label{ux6784ux5efaux7d22ux5f15-1}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{REF}\OperatorTok{=}\NormalTok{\textasciitilde{}/database/Human/Homo\_sapiens.GRCh38.dna.toplevel.fa}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ \textasciitilde{}/src/run/hisat2.mk REF=}\VariableTok{$\{REF\}}\NormalTok{ index}
\end{Highlighting}
\end{Shaded}

\section{比对}\label{ux6bd4ux5bf9-1}

单端数据

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{REF}\OperatorTok{=}\NormalTok{genome/Human/Homo\_sapiens.GRCh38.dna.primary\_assembly.fa}
\VariableTok{HISAT2}\OperatorTok{=}\NormalTok{results/hisat2}
\VariableTok{CLEANDATA}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}
\VariableTok{Run}\OperatorTok{=}\NormalTok{SRR1343245}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ src/run/hisat2.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{        REF=}\VariableTok{$\{REF\}} \DataTypeTok{\textbackslash{}}
\NormalTok{        R1=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/}\VariableTok{$\{Run\}}\NormalTok{\_1.trimmed.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        BAM=}\VariableTok{$\{HISAT2\}}\NormalTok{/\{sample\}.bam }\DataTypeTok{\textbackslash{}}
\NormalTok{        run}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{design}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/design.csv}
\VariableTok{REF}\OperatorTok{=}\NormalTok{genome/Human/Homo\_sapiens.GRCh38.dna.primary\_assembly.fa}
\VariableTok{CLEANDATA}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}
\VariableTok{HISAT2}\OperatorTok{=}\NormalTok{results/hisat2}
                
\FunctionTok{cat} \VariableTok{$\{design\}} \KeywordTok{|} \ExtensionTok{parallel} \AttributeTok{{-}{-}header}\NormalTok{ : }\AttributeTok{{-}{-}colsep}\NormalTok{ , }\DataTypeTok{\textbackslash{}}
\NormalTok{        make }\AttributeTok{{-}f}\NormalTok{ src/run/hisat2.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{        REF=}\VariableTok{$\{REF\}} \DataTypeTok{\textbackslash{}}
\NormalTok{        R1=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/\{Run\}\_1.trimmed.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        BAM=}\VariableTok{$\{HISAT2\}}\NormalTok{/\{sample\}.bam }\DataTypeTok{\textbackslash{}}
\NormalTok{        run}
\end{Highlighting}
\end{Shaded}

双端数据

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{REF}\OperatorTok{=}\NormalTok{genome/Human/Homo\_sapiens.GRCh38.dna.primary\_assembly.fa}
\VariableTok{HISAT2}\OperatorTok{=}\NormalTok{results/hisat2}
\VariableTok{CLEANDATA}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}
\VariableTok{Run}\OperatorTok{=}\NormalTok{SRR1343245}

\FunctionTok{make} \AttributeTok{{-}f}\NormalTok{ src/run/hisat2.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{        REF=}\VariableTok{$\{REF\}} \DataTypeTok{\textbackslash{}}
\NormalTok{        R1=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/}\VariableTok{$\{Run\}}\NormalTok{\_1.trimmed.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        R2=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/}\VariableTok{$\{Run\}}\NormalTok{\_2.trimmed.fastq}
        \VariableTok{BAM}\OperatorTok{=}\VariableTok{$\{HISAT2\}}\NormalTok{/\{sample\}.bam }\DataTypeTok{\textbackslash{}}
        \ExtensionTok{run}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{DATA}\OperatorTok{=}\NormalTok{data}
\VariableTok{design}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/design.csv}
\VariableTok{REF}\OperatorTok{=}\NormalTok{genome/Human/Homo\_sapiens.GRCh38.dna.primary\_assembly.fa}
\VariableTok{CLEANDATA}\OperatorTok{=}\VariableTok{$\{DATA\}}\NormalTok{/reads}
\VariableTok{HISAT2}\OperatorTok{=}\NormalTok{results/hisat2}
                
\FunctionTok{cat} \VariableTok{$\{design\}} \KeywordTok{|} \ExtensionTok{parallel} \AttributeTok{{-}{-}header}\NormalTok{ : }\AttributeTok{{-}{-}colsep}\NormalTok{ , }\DataTypeTok{\textbackslash{}}
\NormalTok{        make }\AttributeTok{{-}f}\NormalTok{ src/run/hisat2.mk }\DataTypeTok{\textbackslash{}}
\NormalTok{        REF=}\VariableTok{$\{REF\}} \DataTypeTok{\textbackslash{}}
\NormalTok{        R1=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/\{Run\}\_1.trimmed.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        R2=}\VariableTok{$\{CLEANDATA\}}\NormalTok{/\{Run\}\_2.trimmed.fastq }\DataTypeTok{\textbackslash{}}
\NormalTok{        BAM=}\VariableTok{$\{HISAT2\}}\NormalTok{/\{sample\}.bam }\DataTypeTok{\textbackslash{}}
\NormalTok{        run}
\end{Highlighting}
\end{Shaded}

\href{src/run/hisat2.mk}{sra.mk}

\bibliography{book.bib,packages.bib}

\backmatter
\printindex

\end{document}
